baseclasses/SolverHist.py

```diff
@@ -0,0 +1,210 @@
+
+    def write(self, data):
+        """
+        Store the values from the latest iteration in the solver history
+        """
+
+        # Figure out if we're storing the history or not
+        if self.storeValues:
+            i = self.latestIter
+        else:
+            i = 0
+
+        self.hist["values"]["Iter"][i] = self.latestIter
+        for key, val in data.items():
+            try:
+                self.hist["values"][key][i] = val
+            except KeyError:
+                Error("Option '%-30s' is not a valid history variable." % (key))
+
+        self.latestIter += 1
+
+    def printHeader(self):
+        """
+        Print the formatted header, used during nonlinear solution process
+        """
+        header = "|"
+        for var in self.varsToPrint:
+            if var in self.hist["nameFormat"]:
+                header += " " + self.hist["nameFormat"][var].format(var) + " |"
+
+        print("+" + "-" * (len(header) - 2) + "+")
+        print(header)
+        print("+" + "-" * (len(header) - 2) + "+")
+
+    def printData(self, ind=None):
+        """
+        Print a selection of the history in the correctly formatted manner, used for printing convergence history during
+        nonlinear solution process
+        """
+        if ind == None:
+            ind = [self.latestIter - 1]
+        if type(ind) is not list:
+            ind = [ind]
+        if not self.storeValues and (len(ind) > 1 or all(i not in [0, -1] for i in ind)):
+            warn(
+                'Previous iterations unavailable as py option "writeConvergenceHist" is not enabled. Printing latest iteration'
+            )
+            ind = [self.latestIter - 1]
+
+        for iter in ind:
+            if iter < self.latestIter:
+                line = "|"
+                for var in self.varsToPrint:
+                    if var in self.hist["nameFormat"]:
+                        if self.hist["varType"][var] in ["float", "int"] and np.isnan(self.hist["values"][var][iter]):
+                            line += " " + self.hist["nameFormat"][var].format("-") + " |"
+                        else:
+                            line += " " + self.hist["valueFormat"][var].format(self.hist["values"][var][iter]) + " |"
+                print(line)
+
+    def clip(self):
+        """
+        Remove the unused entries in the history dict
+        """
+
+        for key in self.hist["values"]:
+            self.hist["values"][key] = self.hist["values"][key][: self.latestIter]
+
+    def save(self, filename):
+        """Postprocess the history and save as a pickled dictionary"""
+
+        import dill as pickle
+
+        base = os.path.splitext(filename)[0]
+        filename = base + ".pkl"
+
+        # Remove the unused entries
+        self.clip()
+
+        # Store the options this problem ran with
+        self.hist["options"] = self.options
+
+        pickle.dump(self.hist, open(filename, "wb"), protocol=pickle.HIGHEST_PROTOCOL)
```

Comment:
from recent experience, I would have the pickle protocol as a `kwarg` here, that would help portability

Reply:
Sure, which protocol works best?

Reply:
`HIGHEST_PROTOCOL` is fine on py3 but for example py2 has `protocol: 4` as highest supported. This is fine with our codes, but in general We should add `protocol = pickle.HIGHEST_PROTOCOL` in case we want to generate a pickle file that is also backwards compatible